[PR1] DEV-14 add Testing page components

jmaralo opened the pull request:
These are the components that will be used in the control section in the testing page, they will be used to control the simulation and to generate disturbances.

jmaralo committed:
DEV-14 add ToggleSwitch

The ToggleSwitch is a simple animated toggle switch.

It will be used in the HIL page next to some input elements

jmaralo committed:
DEV-14 upgrade ToggleButton

add the useToggle to the ToggleButton and update style units to rems.

jmaralo committed:
DEV-14 simplify ToggleSwitch

The ToggleSwitch is now dummer.

jmaralo committed:
DEV-14 add InputTag

The input tag accepts the same elements as an HTML input element.

It wraps an input inside a border with a label

jmaralo committed:
DEV-14 add ToggleInput

ToggleInput enables or disables an InputLabel based on a ToggleSwitch.

jmaralo committed:
DEV-14 add final components

Added some additional components

Some of the previous ones have been slightly improved

lipezaballa committed:
DEV-14: css changes

lipezaballa committed:
DEV-14: TestControls

lipezaballa committed:
DEV-14: Minor changes

lipezaballa committed:
DEV-14: Minor changes

jmaralo committed:
Merge branch 'develop' into features/DEV-14

Losina24 commented on src/components/ToggleButton/ToggleButton.tsx:
```diff
@@ -1,31 +1,39 @@
+        }, [isOn])
     }
 
+    let onClickFlip = (_: any) => {
```
(_:any) => flip()

Losina24 commented on src/components/ToggleInput/ToggleInput.tsx:
```diff
@@ -0,0 +1,28 @@
+export function ToggleInput({onToggle, disabled, ...inputProps}: Props) {
+    const [isOn, flip] = useToggle();
+
+    if (onToggle) {
```
you should try to reduce the indentations as much as possible

Losina24 commented on src/pages/TestingPage/TestControls/TestControls.tsx:
```diff
@@ -0,0 +1,103 @@
+                        <ToggleButton label="brake" icon={<BreakIcon />} />
+                    </div>
+                    <div className={style.inputWrapper}>
+                        <ToggleInput
```
Commented on slack

jmaralo committed:
DEV-14 reduce identations

Two small changes to reduce the indentation and make code cleaner

jmaralo committed:
Merge branch 'features/DEV-14' of https://github.com/HyperloopUPV-H8/control-front into features/DEV-14

smorsua commented on src/components/ButtonTag/ButtonTag.module.scss:
```diff
@@ -0,0 +1,65 @@
+    height: fit-content;
+    width: 100%;
+
+    p {
```
Dont use element selectors, they affect the whole website

jmaralo replied:
in this case I don't think it applies to the whole website, as it is only if it is a child of a buttonTagWrapper

smorsua commented on src/components/ButtonTag/ButtonTag.tsx:
```diff
@@ -0,0 +1,19 @@
+    return (
+        <label className={style.buttonTagWrapper}>
+            <button {...buttonProps}>{icon}</button>
+            {label != null ? <p>{label}</p> : null}
```
Do `{label && <p>{label}</p>}` instead

smorsua commented on src/components/InputTag/InputTag.module.scss:
```diff
@@ -0,0 +1,63 @@
+    -moz-appearance: textfield;
+}
+
+legend {
```
Dont use element selector

smorsua commented on src/components/InputTag/InputTag.module.scss:
```diff
@@ -0,0 +1,63 @@
+    flex-grow: 1;
+    height: 2.8rem;
+
+    legend {
```
Here i dont mind because is inside another class

smorsua commented on src/components/PlayButton/PlayButton.tsx:
```diff
@@ -0,0 +1,61 @@
+import { FaPlay, FaPause, FaStop } from "react-icons/fa";
+import style from "./PlayButton.module.scss";
+
+type PlayButtonTypeAttribute = "play" | "pause" | "stop" | "disabled";
```
Change to ButtonType = keyof typeof buttonVariants

jmaralo replied:
That either creates a cyclic reference, or forces me to change the type of playButtonVariants to something less meaningful, I prefer this approach, but I would like to know if you can think of any way (setting playButtonVariants key type to string and setting it as const didn't work)

smorsua commented on src/components/PlayButton/PlayButton.tsx:
```diff
@@ -0,0 +1,61 @@
+    colorClass: string;
+};
+
+const playButtonVariants: Record<PlayButtonTypeAttribute, Variant> = {
```
Rename to buttonVariants

jmaralo replied:
I feel playButtonVariants is more descriptive, why do you want to change the name?

smorsua replied:
Because play is one of the types, you also have pause, stop, disabled

smorsua commented on src/components/ToggleSwitch/ToggleSwitch.module.scss:
```diff
@@ -0,0 +1,48 @@
+$switch-aspect-ratio: 2;
+$switch-padding: 0.25rem;
+
+label {
```
Dont use element selectors

smorsua commented on src/components/ToggleSwitch/ToggleSwitch.module.scss:
```diff
@@ -0,0 +1,48 @@
+
+    transition: padding-left 100ms ease-in, background-color 100ms ease-in;
+
+    div {
```
Same

smorsua commented on src/hooks/useToggle.ts:
```diff
@@ -0,0 +1,9 @@
+import { useState } from "react"
+
+export function useToggle() {
+    const [isToggeled, setIsToggeled] = useState<boolean>(false)
```
Remove type argument (it infers it)

smorsua commented on src/components/ToggleSwitch/ToggleSwitch.tsx:
```diff
@@ -0,0 +1,16 @@
+    flip: () => void,
+}
+
+export function ToggleSwitch({ isOn, flip }: Props) {
```
Put the useToggle here and accept a callback a a prop that is called when the toggle is switched. The callback should receive the isOn value

smorsua reviewed (changes requested):
Good overall, get back to me when you apply the changes

jmaralo committed:
DEV-14 remove global css selectors and move logic

Changes to the CSS so no element selector is used outside a wrapper

Move the toggle logic of ToggleInput to ToggleSwitch

jmaralo committed:
DEV-14 infer PlayButtonTypeAttribute

smorsua commented on src/components/ToggleButton/ToggleButton.tsx:
```diff
@@ -1,31 +1,38 @@
-            props.onToggle(target, next)
-            return next
-        })
+    if (onToggle) {
```
Conditional hooks are not allowed, do

`useEffect(()=>{
onToggle?.() // This syntax is like conditional calling the function if is not undefined
}, [isOn])`

smorsua commented on src/components/ToggleButton/ToggleButton.tsx:
```diff
@@ -1,31 +1,38 @@
 
-export function ToggleButton(props: Props) {
-    const [isToggeled, setIsToggeled] = useState(false)
+export function ToggleButton({ label, icon, onToggle, onClick, ...buttonProps }: Props) {
```
Remove the onClick, if the consumer of the component doesnt want the isOn value, they just dont provide a parameter in their callback

jmaralo committed:
DEV-14 rename variables and remove unnecessary props

Rename some types and variables in the PlayButton component

Remove the onClick prop for the ToggleButton component

jmaralo committed:
DEV-14 remove conditionally executing hook

The useEffect of ToggleButton was inside a conditional

smorsua commented on src/components/ToggleSwitch/ToggleSwitch.tsx:
```diff
@@ -0,0 +1,22 @@
+
+    const onClick = () => {
+        flip();
+        onToggle(isOn);
```
You have to use the useEffect like in the ToggleButton because isOn will be the old value instead of the new one.

jmaralo committed:
DEV-14 fix ToggleSwitch hook bug

onToggle is called inside useEffect, after isOn is updated
